lib/kdl/include/kdl/transform_range.h

```diff
@@ -0,0 +1,148 @@
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef TRENCHBROOM_TRANSFORM_RANGE_H
+#define TRENCHBROOM_TRANSFORM_RANGE_H
+
+#include <type_traits>
+
+namespace kdl {
+    /**
+     * Wraps an iterator and applies a transformation function and returns its result every time it is dereferenced.
+     *
+     * @tparam I the type of the wrapped iterator
+     * @tparam Transform the type of the transformation to apply
+     */
+    template <typename I, typename Transform>
+    class transform_iterator {
+    public:
+        using iterator_category = typename I::iterator_category;
+        using value_type = std::invoke_result_t<Transform, typename I::value_type>;
+        using difference_type = typename I::difference_type;
+        using pointer = value_type*;
+        using reference = value_type&;
+    private:
+        I m_iter;
+        Transform m_transform;
+    public:
+        /**
+         * Creates a new transform iterator with the given wrapped iterator and transformation function.
+         *
+         * @param iter the wrapped iterator
+         * @param transform the transformation function, must be movable
+         */
+        transform_iterator(I iter, Transform transform) :
+            m_iter(iter),
+            m_transform(std::move(transform)) {}
+
+        bool operator<(const transform_iterator& other) const { return m_iter <  other.m_iter; }
+        bool operator>(const transform_iterator& other) const { return m_iter >  other.m_iter; }
+        bool operator==(const transform_iterator& other) const { return m_iter == other.m_iter; }
+        bool operator!=(const transform_iterator& other) const { return m_iter != other.m_iter; }
+
+        // prefix
+        transform_iterator& operator++() { ++m_iter; return *this; }
+        transform_iterator& operator--() { --m_iter; return *this; }
+
+        // postfix
+        transform_iterator operator++(int) { transform_iterator result(*this); ++m_iter; return result; }
+        transform_iterator operator--(int) { transform_iterator result(*this); --m_iter; return result; }
+
+        value_type operator*()  { return m_transform(*m_iter); }
+    };
+
+    /**
+     * Adapts a collection with the given transformation function. Adaptation means that the collection produces
```

Comment:
I'd describe it like:
"Wraps a const reference to a collection, and provides transform_iterator's which expose the underlying collection's elements transformed by Transform on every iterator dereference."

Reply:
done